File: haskell/egcd.hs
--
-- egcd m n = (s,t,d) such that
-- d = gcd(m,n) = sm + tn
--
egcd :: Int -> Int -> (Int,Int,Int)
egcd = egcdaux 1 0 0 1

--
-- egcdaux 1 0 0 1 m n = (s,t,d) such that
-- d = gcd(m,n) = sm + tn
--
egcdaux :: Int -> Int -> Int -> Int -> Int -> Int -> (Int,Int,Int)
egcdaux a' b' a b c d =
   let
       q = div c d
       r = rem c d
   in
      case r of 0 -> (a,b,d)
                _ -> egcdaux a b (a' - q * a) (b' - q * b) d r

File: erlang/egcd.erl
-module (egcd).
-export ([egcd/2]).

%  egcd (m,n) = (s,t,d) such that
%    d = gcd(m,n) and
%    sm + tn = d
egcd(M,N) ->
    egcdaux (1,0,0,1,M,N).

%  initialize with (1,0,0,1,m,n) such that
%    a'm + b'n = c   and
%    am  + bn  = d
%  then egcdaux (1,0,0,1,m,n) = (s,t,d) such that
%    d = gcd(m,n) and
%    sm + tn = d
egcdaux(A1,B1,A,B,C,D) ->
    case C rem D of
        0 -> {A,B,D};
        R ->
            Q = C div D,
            egcdaux (A,B,A1 - Q*A, B1 - Q*B,D,R)
    end.
